### PR TITLE
Add libfuzzer compatible fuzz harness

### PR DESCRIPTION
**Pull request by silvergasp**
As discussed in #4088 the fuzz_simplify harness is an easy candidate to begin support for coverage-driven feedback required for high-performance fuzz testing.

### CODE REVIEW

**Comment by silvergasp** [comment]
I wasn't sure on how best to manage the Cmake build files for fuzz tests, so I'm certainly open to suggestions on some of the changes here :)

**Comment by alexreinking** [comment]
Because of the special toolchain requirements, I think this test should be moved to `test/fuzz/simplify.cpp`. That folder should get its own `CMakeLists.txt` containing the following:

```
tests(GROUPS fuzz
      SOURCES
      simplify.cpp
      )
target_compile_options(fuzz_simplify PRIVATE -fsanitize=fuzzer,address)
target_link_options(fuzz_simplify PRIVATE -fsanitize=fuzzer,address)
```

The `test/CMakeLists.txt` should be modified to contain the following:

```
# near the top:
include(CheckCXXCompilerFlag)

# toward the bottom
check_cxx_compiler_flag("-fsanitize=fuzzer,address" HAS_FUZZ_FLAGS)
cmake_dependent_option(
    WITH_TEST_FUZZ "Build fuzz tests" ON
    HAS_FUZZ_FLAGS OFF
)
```

**Comment by silvergasp** [comment]
Thanks for the quick response, I've made the suggested changes and also added some of the precursors for multi-fuzzer engine builds for oss-fuzz:)

File: test/CMakeLists.txt
Comment (steven-johnson) [inline test/CMakeLists.txt:66]: What compiler(s) support this flag? We need documentation on this somewhere, at least in a comment here or elsewhere.

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [inline test/fuzz/simplify.cpp:3]: What happens if libfuzzer isn't installed on the build machine -- will this test be able to compile?

Comment (silvergasp) [reply test/fuzz/simplify.cpp:3]: Libfuzzer is shipped with clang as part of the llvm project as clang/llvm are already dependencies of this project I don't think that this will be an issue. That specific [header file](https://github.com/llvm/llvm-project/blob/main/compiler-rt/include/fuzzer/FuzzedDataProvider.h) is sort of fuzzer adjacent, and is more of a helper for transforming raw data into structured data. But the header itself is not required to write a fuzzer.

File: test/CMakeLists.txt
Comment (silvergasp) [reply test/CMakeLists.txt:66]: As far as I'm aware if we were to break this up into two flags;
- `-fsanitize=fuzzer` supported by clang+msvc
- `-fsanitize=address` aka ASAN support by clang+gcc+msvc

Note that the actual flags for msvc are slightly different e.g. `\fsanitize` rather than `-fsanitize`

Comment (silvergasp) [reply test/CMakeLists.txt:66]: I'll add a comment here :)

File: test/fuzz/CMakeLists.txt
Comment (silvergasp) [reply test/fuzz/CMakeLists.txt:10]: The `LIB_FUZZING_ENGINE` is a somewhat well-known variable. It's used to allow for overriding the fuzzing engine entirely. So the goal here is to have libfuzzer+address-sanitizer be the default but then if LIB_FUZZING_ENGINE is declared you can override the fuzzing engine to one of;
- [Centipede](https://github.com/google/centipede)
- [Hongfuzz](https://github.com/google/honggfuzz)
- [AFL++](https://github.com/AFLplusplus/AFLplusplus)
- etc.

Using this variable is also a [precursor to integrating with OSS-fuzz](https://google.github.io/oss-fuzz/advanced-topics/ideal-integration/#build-support). So while I could rename it to `HALIDE_LIB_FUZZING_ENGINE` it would probably add some unnecessary indirection further down the line.

**Comment by silvergasp** [comment]
Looks like I've got at least a couple of formatting things to fix. 

On a different note, I've got a corresponding [draft PR to integrate with OSS-fuzz](https://github.com/google/oss-fuzz/pull/10107). A couple of things I'll need for that integration to go through;
- A primary contact email address (associated with a google/gmail account).
- A list of email addresses that you want to have CC'd into automatic bug reports from oss-fuzz.
- A contributor here with write access to Halide to comment mentioning that you approve of integration with oss-fuzz.

If you are happy for me to do so, I can pull the email addresses from the git log, it's worth noting however that these emails will be listed in plaintext in the oss-fuzz repo.

File: test/CMakeLists.txt
Comment (steven-johnson) [reply test/CMakeLists.txt:66]: > \fsanitize

...flags starting with a backslash, truly the end times are nigh :-)

File: test/fuzz/CMakeLists.txt
Comment (steven-johnson) [reply test/fuzz/CMakeLists.txt:10]: OK, cool. Please try to integrate these comments into the CMake file or whatnot so that they are captured in the source rather than just here in the PR -- it will make our lives easier in the future.

File: test/CMakeLists.txt
Comment (alexreinking) [reply test/CMakeLists.txt:66]: > > \fsanitize
> 
> ...flags starting with a backslash, truly the end times are nigh :-)

No, the MSVC flag starts with a `/` not a `\`.

Comment (alexreinking) [reply test/CMakeLists.txt:66]: But AFAIK, MSVC understands `-` for the sake of compatibility, too.

File: test/fuzz/CMakeLists.txt
Comment (alexreinking) [inline test/fuzz/CMakeLists.txt:10]: Please do not use environment variables this way in CMake. Unlike with autotools, the user does not control when the build system is regenerated; the build system itself does according to CMake's analysis of the build file dependencies, which includes CMake's own implementation. This will cause vexing behavior when a developer upgrades CMake. Imagine:

> I upgraded CMake and now the fuzz tests aren't building! Why? Oops, I didn't set `LIB_FUZZING_ENGINE` in Bash again.

Thus, cache variables should always dominate. To initialize a cache variable from the environment, write:

```suggestion
set(LIB_FUZZING_ENGINE "$ENV{LIB_FUZZING_ENGINE}"
    CACHE STRING "Compiler flags necessary to use libFuzzer instrumentation")

if (LIB_FUZZING_ENGINE)
```

File: test/CMakeLists.txt
Comment (alexreinking) [reply test/CMakeLists.txt:66]: > What compiler(s) support this flag? We need documentation on this somewhere, at least in a comment here or elsewhere.

Part of the point of checking is that it's less necessary to know the complete matrix of compiler vendors and versions. We could add some `message(VERBOSE)`s for debugging when detection fails.

**Review by alexreinking — CHANGES_REQUESTED** [review]

**Comment by steven-johnson** [comment]
> Part of the point of checking is that it's less necessary to know the complete matrix of compiler vendors and versions

No, but when it inevitably fails somewhere, having comments and/or messages to help explain would be nice

File: test/fuzz/CMakeLists.txt
Comment (silvergasp) [reply test/fuzz/CMakeLists.txt:10]: Cool thanks for the suggestion. I'll add that in. I'm definitely a cmake noob, so I appreciate the guidance.

Comment (alexreinking) [reply test/fuzz/CMakeLists.txt:10]: > I'm definitely a cmake noob,

For your own sanity, stay that way 😆

Comment (silvergasp) [reply test/fuzz/CMakeLists.txt]: > > I'm definitely a cmake noob,
> 
> For your own sanity, stay that way 😆

lol yeah, I find cmake in general just... baffling

**Comment by silvergasp** [comment]
Ok so I think there is a bit of [a bug with cmakes `CHECK_CXX_COMPILER_FLAG`](https://cmake.cmake.narkive.com/faC8opxU/check-cxx-compiler-flag-usage) as it doesn't seem to support checking sanitiser flags. Something to do with it needing those sanitizer flags during linking as well, so the check fails. I'm going to have a bit of a go at using CheckCXXSourceCompiles instead as it seems to give a bit more control over how/what flags are used during compiling/linking.

File: test/CMakeLists.txt
Comment (alexreinking) [reply test/CMakeLists.txt:85]: No. See the documentation here: https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html

If `HAS_FUZZ_FLAGS` is false, then `WITH_TEST_FUZZ` must be false, too.

File: test/CMakeLists.txt
Comment (alexreinking) [inline test/CMakeLists.txt:76]: I'd prefer to see this as a multi-line string
```suggestion
check_cxx_source_compiles([[
#include <cstdint>
extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, std::size_t Size) {
  return 0;
}
]] HAS_FUZZ_FLAGS)
```

File: test/CMakeLists.txt
Comment (silvergasp) [reply test/CMakeLists.txt:90]: Cool, yeah I've switched back in the latest revision, it looks like it was more a problem with the flag check than how the dependant option was configured.

**Comment by silvergasp** [comment]
I think I've covered most things here. Is [this build failure related to my changes](https://buildbot.halide-lang.org/master/#/builders/52/builds/330). The error is;
```
The following tests FAILED:
	243 - correctness_multi_way_select (Failed)
Errors while running CTest
program finished with exit code 8
elapsedTime=578.168036
```

File: test/CMakeLists.txt
Comment (silvergasp) [reply test/CMakeLists.txt]: Done in https://github.com/halide/Halide/pull/7512/files#diff-33394812ba204689144fd2f80832db83853ba1cb32403edb4e15fe4893e675fdR84-R86

**Review by steven-johnson — APPROVED** [review]

**Comment by steven-johnson** [comment]
LGTM, waiting on @alexreinking approval

**Review by alexreinking — APPROVED** [review]
This is great! Thanks for the contribution!

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:28]: TODO: change as per #7546

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:49]: Looks fine as per #7546

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:57]: Should be `fdp.ConsumeIntegralInRange<int>(0, fuzz_var_count)` as per #7546 NOTE: drop the `+ 1`

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:64]: Should be `fdp.ConsumeIntegralInRange<int>(-128, 127)` as per #7546

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:66]: Should be `fdp.ConsumeIntegralInRange<int>(-RAND_MAX / 2, RAND_MAX / 2 -1)` as per #7546

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:102]: Should be `fdp.ConsumeIntegralInRange<int>(0, op_count-1)` as per #7546

File: test/fuzz/simplify.cpp
Comment (silvergasp) [inline test/fuzz/simplify.cpp:142]: Should be `fdp.ConsumeIntegralInRange<int>(0, op_count-1)`

**Review by silvergasp — COMMENTED** [review]
@steven-johnson would you mind double checking these? I'll have a PR up shortly with these changes.

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [inline test/fuzz/simplify.cpp:31]: Hadn't noticed this call existed -- perhaps we should aggressively use it in other cases too where possible (e.g. line 47, line 102)

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [reply test/fuzz/simplify.cpp:57]: The semantics of fuzz_var_count aren't clear to me -- in various for loops we do `for (int i = 0; i < fuzz_var_count; i++)` which implies that we need to limit the range to max=fuzz_var_count-1?

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [reply test/fuzz/simplify.cpp:66]: LGTM

Comment (steven-johnson) [reply test/fuzz/simplify.cpp:66]: ...except, RAND_MAX applies to the `rand()` call in stdlib; does libfuzzer also make the same promise about range?

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [reply test/fuzz/simplify.cpp:102]: or better yet, use PickValueInArray()?

Comment (silvergasp) [reply test/fuzz/simplify.cpp:102]: True that would be better

File: test/fuzz/simplify.cpp
Comment (steven-johnson) [reply test/fuzz/simplify.cpp:142]: (I realize the fragility was pre-existing before your changes, but improving the resilience here seems easy and worthwhile while we are here)

File: test/fuzz/simplify.cpp
Comment (silvergasp) [reply test/fuzz/simplify.cpp:57]: Yeah looking at this again the `var` variable seems to only get used once on line 56, so I could probably just get rid of it and replace the condition on line 56 with `fdp.ConsumeBool`.

File: test/fuzz/simplify.cpp
Comment (silvergasp) [reply test/fuzz/simplify.cpp:142]: Yeah, looking at this again a second time I'm struggling to understand what is going on here. Where did [this `5`](https://github.com/halide/Halide/pull/7512/files#diff-0bb69bc3a0270944620451c6da8561235b4d1062aa5acea5357e3aa29f99e451R140) come from.

File: test/fuzz/simplify.cpp
Comment (silvergasp) [reply test/fuzz/simplify.cpp:57]: I'm guessing it was +1 in the beginning so that the condition on line 56 was sometimes true/false. In any case it seems like a really roundabout way of doing it. I'm just going to replace it with the consumebool method.

**Pull request merged**
